### PR TITLE
[api] Add BCS as an alternative to JSON with similar but slightly different types

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
For performance reasons, BCS is a lot faster.  If we don't convert the structs in the middle, then we save time, plus JSON is just generally slower.

APIs that are supported in this PR:
* get account
* get ledger info (was already supported)
* get block by height / version
* get account resources
* get account modules
* get account transactions
* get transactions
* simulate transaction
* submit transaction
* get transaction by hash
* get transaction by version

APIs that are not supported in this PR:
* get events by key
* get events by handle
* get account resource
* get account module
* get table item

### Test Plan
There is one E2E test that handles all of these, comparing them to the ones that already exist.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3110)

### CODE REVIEW

File: api/src/accounts.rs [hunk 1]
```diff
@@ -164,50 +165,88 @@ impl Account {
...
-            BasicResponseStatus::Ok,
-            accept_type,
-        ))
+        match accept_type {
+            AcceptType::Json => BasicResponse::try_from_json((
+                account_data,
+                &self.latest_ledger_info,
+                BasicResponseStatus::Ok,
+            )),
+            AcceptType::Bcs => BasicResponse::try_from_encoded((
```
Comment (davidiw) [inline api/src/accounts.rs:174]: will this break @jjleng's BCS API?
if so we should hold off on the wallet using this

File: testsuite/smoke-test/src/rest_api.rs [hunk 2]
```diff
@@ -51,3 +59,233 @@ async fn test_basic_client() {
 
     info.client().get_transactions(None, None).await.unwrap();
 }
+
+#[tokio::test]
+async fn test_bcs() {
```
Comment (davidiw) [inline testsuite/smoke-test/src/rest_api.rs:64]: what's the advantage of this over the REST api tests?

**Review by davidiw — APPROVED** [review]

File: testsuite/smoke-test/src/rest_api.rs (hunk 2 above)
Comment (gregnazario) [reply testsuite/smoke-test/src/rest_api.rs:64]: The advantage here is I properly tested the REST client, which we had no tests for.

It's actually the reason we missed version -> ledger_version

File: api/src/accounts.rs (hunk 1 above)
Comment (gregnazario) [reply api/src/accounts.rs:174]: I've got to look at what it uses.  This is exactly why I said earlier we shouldn't allow responses on the BCS API unless it's been properly completed.

**Comment by jjleng** [comment]
What about defining a generalized ABI standard that can express functions and resource/struct layouts? And each endpoint returns the BCS payloads plus the hash code of the ABI layout of the payload. Clients can download the ABIs and cache them. Upon receiving the BCS result, clients can get the ABI by ABI hash from the local cache. This way clients can decode BCS results.

Note, the ABI standard will be so useful for other use cases. Clients can even generate code for smart contract resources.

**Pull request merged**